### DISCUSSION TITLE
[Feature Request] 当使用其他DNS时路由器自身如何劫持DNS请求？

### DISCUSSION THREAD

**Discussion by cuminlo** [nikkinikki-org/OpenWrt-nikki#242] [Q&A]
### 确认

- [X] 我已经查看过Wiki，没有找到我想要的功能
- [x] 我已经搜索过Issue，没有找到相同/类似的功能请求

### 功能描述

感谢大佬提供的插件，简单好用！

我的使用场景是，`DNS` 请求到 `AdGuardHome`，分流之后国外域名到 `OpenWrt-mihomo`
- 开启绕过
```
DNS -> AdGuardHome -> OpenWrt-mihomo
```
这样就没法使用DNS劫持功能，于是我使用端口转发功能实现，如图配置（5553端口是`AdGuardHome`）
![Xnip2024-10-16_16-26-31](https://github.com/user-attachments/assets/db2a1380-5ff5-43a8-a775-3f2274ea35cd)
这样配置之后，局域网内的请求确实按照预期经过了 `AdGuardHome`，分流和绕过功能我目测是正常的。

但是路由器自身的`DNS`请求没有经过 `AdGuardHome`, 以前我使用 `OpenClash` 也遇到过这个问题，它那边提供的方案是允许启动后添加用户自己的脚本，我在脚本里面实现劫持让路由器自身也能代理。

### 如何实现

是否能实现类似的功能，或者其他更好的方式，或者能够提供相关命令我自己手动执行也行，感激不尽！

**Comment by zutzo** [comment]
？，你当 dnsmasq 是摆设吗
将 AGH 设置成 dnsmasq 上游

**Comment by morytyann** [comment]
这不是一个功能请求，你应该开在discussion里。

解决办法如上，即`将 AGH 设置成 dnsmasq 上游`，如图
![{DDB8B358-C75D-4CE9-A244-C0743080C0E1}](https://github.com/user-attachments/assets/3f4003ad-5598-48b0-9f6a-9e4d6a3af9b9)

最后的链路是`客户端 -> dnsmasq -> AGH -> mihomo`。

**Reply by morytyann** [reply]
不过会导致AGH的客户端统计功能失效，想解决这个问题，你可以装一个luci-app-adguardhome，模式选劫持DNS请求到AGH。

**Comment by cuminlo** [comment]
> ？，你当 dnsmasq 是摆设吗 将 AGH 设置成 dnsmasq 上游

@zutzo  @morytyann 感谢，我刚试了一下确实可以的。

最开始我是使用的上游方式，但是`AGH`会全部显示 `127.0.0.1`，所以我才改成了端口转发。

我以为两者不能并行，所以启用端口转发后就把 `AGH `的劫持都关闭了，我现在两者都开启了，目前看日志是正常的。

**Comment by cuminlo** [comment]
最后的链路应该是这样的吧...

```
局域网客户端 -> AdGuardHome -> mihomo
```

```
路由器自身-> dnsmasq -> AdGuardHome -> mihomo
```

**Reply by morytyann** [reply]
是的，你这是端口转发来处理客户端，上游来处理路由器自身了🤣

其实最好的办法还是装个luci-app-adguardhome，然后用它来劫持，这样两个行为就一致了

**Reply by zackfan** [reply]
> 是的，你这是端口转发来处理客户端，上游来处理路由器自身了🤣
> 
> 其实最好的办法还是装个luci-app-adguardhome，然后用它来劫持，这样两个行为就一致了

@morytyann 我使用最新版的nikki，流程也是ad->nikki，关闭了dnsmasq的重定向，把53重定向到adguardhome的端口，正常上网没问题，但是路由器更新软件包失败，这是为啥呢？

**Reply by morytyann** [reply]
删除`代理配置 -> 绕过 -> 绕过资源控制组`里的`uhttpd`和`nginx`应该就可以了。
